### ISSUE TITLE
How to print training time

### ISSUE THREAD

**Issue by wsnooker** [OpenNMT/OpenNMT#116]
Want to compare training time with different model structure & parameters.

And now, the training takes 6 days per epoch with 60 million sentence pairs in a single k40 nvidia GPU.

Model parameters:
wordvec: 500   rnn hidden: 1000   batchsize: 64   optim: sgd    single rnn layer

Having any advice for speeding up?
How about reducing rnn hidden or increasing bathsize  or using adam/adadelta for fast convergence?

**Comment by guillaumekln** [comment]
To compare training speed, `Source tokens/s` is usually a good indicator.

Otherwise, you could simply rely on the times displayed in the log (in the format `[02/08/17 17:19:26 INFO]`) to compare elapsed time during the training.

**Comment by wsnooker** [comment]
I use the release 0.2 version and could not find the times displayed in the log. Is it added in release 0.3 version?

**Comment by guillaumekln** [comment]
Yes, advanced logs were added in v0.3.0. You can also deduce training time with saved models by comparing their date.

---

There are no magical tricks to speedup the training without changing the model configuration. However, we are working on several features that could help: CUDNN support, adaptive softmax and half float support.

Increasing the batch size does make the training faster but a too large value can impact convergence (values above 128 seem to not work well). If you use 128 for example, it will also double the GPU memory usage.

I would not recommend using another optimizers. We did not always have good experience with them.

**Comment by wsnooker** [comment]
How to get `Source tokens/s` ?

How about the convergence and translation result when setting bath size to 128 compared to default ?

**Comment by guillaumekln** [comment]
`Source tokens/s` is displayed in the training logs (fourth field):

```
Epoch 1 ; Iteration 1/183 ; Learning rate 1.0000 ; Source tokens/s 9 ; Perplexity 36920.62
```

We did a test when implementing data parallelism and using a batch size of 128 yields similar (yet slightly higher) perplexity on the validation set.

**Comment by wsnooker** [comment]
I could not get the idea of data parallelism and how could I set it ?

**Comment by guillaumekln** [comment]
Take a look at the guide: http://opennmt.net/Advanced/#multi-gpu-training

**Comment by wsnooker** [comment]
ok, thanks for your kindly help.
